certora/specs/SocialRecoveryModule.spec

```diff
@@ -1,26 +1,131 @@
+    function threshold(address) external returns (uint256) envfree;
+
+    // Guardian Storage Functions
+    function guardianStorageContract.countGuardians(address) external returns (uint256) envfree;
+
     // Safe Functions
-    function safeContract.isModuleEnabled(address) external returns (bool) envfree;
+    function safeContract.isModuleEnabled(address module) external returns (bool) envfree;
+    function safeContract.isOwner(address owner) external returns (bool) envfree;
+
+    // Wildcard Functions (Because of use of ISafe interface in Social Recovery Module)
+    function _.isModuleEnabled(address module) external => safeIsModuleEnabled(module) expect bool ALL;
+    function _.isOwner(address owner) external => safeCallerCheckIsOwner(calledContract, owner) expect bool ALL; // `calledContract` is a special variable.
+}
+
+// A summary function to check if a module is enabled in the Safe contract.
+function safeIsModuleEnabled(address module) returns bool {
+    return safeContract.isModuleEnabled(module);
+}
+
+// A summary function to check if an address is an owner in the Safe contract.
+function safeCallerCheckIsOwner(address calledContractAddress, address owner) returns bool {
+    assert calledContractAddress == safeContract; // This checks that the caller is Safe Smart Account.
+    return safeContract.isOwner(owner);
 }
 
-// A setup function that requires Safe contract to enabled the Social Recovery
-// Module.
+// A setup function that requires Safe contract to enabled the Social Recovery Module.
 function requireSocialRecoveryModuleEnabled() {
     require(safeContract.isModuleEnabled(currentContract));
 }
 
-// This is a dummy rule to verify the Safe Contract Setup with the Social Recovery
-// Module is working as intended.
-rule recoveryModuleCanBeDisabled {
-    env e;
-    address prevModule;
+// This integrity rule verifies that the addGuardianWithThreshold(...) works as expected.
+rule addGuardianWorksAsExpected(env e, address guardian, uint256 threshold, address otherGuardian) {
+
+    // This can be removed once linked list invariant is implemented.
+    require guardianStorageContract.entries[safeContract].count > 0 => guardianStorageContract.entries[safeContract].guardians[SENTINEL()] != 0; // Sentinel should not be zero if there are guardians.
+
+    require guardian != otherGuardian;
+
+    bool otherWasGuardian = currentContract.isGuardian(safeContract, otherGuardian);
+
+    currentContract.addGuardianWithThreshold(e, safeContract, guardian, threshold);
+
+    assert safeContract.isModuleEnabled(currentContract);
+    assert e.msg.sender == safeContract;
+    assert currentContract.isGuardian(safeContract, guardian);
+    assert otherWasGuardian == currentContract.isGuardian(safeContract, otherGuardian);
+}
+
+// This integrity rule verifies that the guardian can always be added considering ideal conditions.
+rule guardianCanAlwaysBeAdded(env e, address guardian, uint256 threshold, address otherGuardian) {
```

Comment:
the `wasOtherGuardian == isGuardian(...)` is already checked in the integrity rule (which proves that it doesn't change for any `addGuardian` call that doesn't revert). IMO it adds unnecessary noise to this rule and doesn't prove any additional guarantees.

Reply:
Removed 👍🏾